fhir-server/src/main/java/au/csiro/pathling/fhirpath/element/QuantityPath.java

```diff
@@ -30,4 +46,102 @@ protected QuantityPath(@Nonnull final String expression, @Nonnull final Dataset<
+    return COMPARABLE_TYPES.contains(type);
+  }
+
+  @Nonnull
+  @Override
+  public Column getNumericValueColumn() {
+    return getValueColumn().getField(QuantityEncoding.CANONICALIZED_VALUE_COLUMN);
+  }
+
+  @Nonnull
+  @Override
+  public Column getNumericContextColumn() {
+    return getValueColumn();
+  }
+
+  @Nonnull
+  @Override
+  public Function<Numeric, NonLiteralPath> getMathOperation(@Nonnull final MathOperation operation,
+      @Nonnull final String expression, @Nonnull final Dataset<Row> dataset) {
+    return buildMathOperation(this, operation, expression, dataset, getFhirType());
+  }
+
+  @Nonnull
+  public static Function<Numeric, NonLiteralPath> buildMathOperation(@Nonnull final Numeric source,
+      @Nonnull final MathOperation operation, @Nonnull final String expression,
+      @Nonnull final Dataset<Row> dataset, @Nonnull final FHIRDefinedType fhirType) {
+    return target -> {
+      final Column sourceComparable = source.getNumericValueColumn();
+      final Column sourceContext = source.getNumericContextColumn();
+      final Column targetContext = target.getNumericContextColumn();
+      final Column resultColumn = operation.getSparkFunction()
+          .apply(sourceComparable, target.getNumericValueColumn());
+      final Column sourceCanonicalizedCode = sourceContext.getField(
+          QuantityEncoding.CANONICALIZED_CODE_COLUMN);
+      final Column targetCanonicalizedCode = targetContext.getField(
+          QuantityEncoding.CANONICALIZED_CODE_COLUMN);
+      final Column resultStruct = struct(
+          sourceContext.getField("id").as("id"),
+          resultColumn.as("value"),
+          lit(null).as("value_scale"),
+          sourceContext.getField("comparator").as("comparator"),
+          sourceCanonicalizedCode.as("unit"),
+          sourceContext.getField("system").as("system"),
+          sourceCanonicalizedCode.as("code"),
+          resultColumn.as(QuantityEncoding.CANONICALIZED_VALUE_COLUMN),
+          sourceCanonicalizedCode.as(QuantityEncoding.CANONICALIZED_CODE_COLUMN),
+          sourceContext.getField("_fid").as("_fid")
+      );
+      final Column resultQuantityColumn =
+          when(sourceCanonicalizedCode.equalTo(targetCanonicalizedCode), resultStruct)
+              .otherwise(null);
+
+      final Column idColumn = source.getIdColumn();
+      final Optional<Column> eidColumn = findEidColumn(source, target);
+      final Optional<Column> thisColumn = findThisColumn(source, target);
+
+      switch (operation) {
+        case ADDITION:
+        case SUBTRACTION:
+        case MULTIPLICATION:
+        case DIVISION:
+          return ElementPath
+              .build(expression, dataset, idColumn, eidColumn, resultQuantityColumn, true,
+                  Optional.empty(),
+                  thisColumn, fhirType);
+        default:
+          throw new AssertionError("Unsupported math operation encountered: " + operation);
+      }
+    };
```

Comment:
Perhaps can be delegated to a counterpart of Comparator for arithmetic operations.